engine/src/main/java/org/hibernate/search/spi/SearchIntegratorBuilder.java

```diff
@@ -403,6 +407,35 @@ else if ( metadataProvider.containsSearchMetadata( mappedClass ) ) {
 	}
 
 	/**
+	 * Get the filter definitions from the super classes that are not annotated with @Indexed
+	 */
+	private void bindFilterDefsInSuperClasses(final Map<XClass, Class<?>> classMappings) {
+		Set<XClass> superClasses = new HashSet<XClass>();
+		for ( Map.Entry<XClass, Class<?>> mapping : classMappings.entrySet() ) {
+			saveSuperClasses( superClasses, mapping.getKey() );
+		}
+		for ( XClass xClass : superClasses ) {
+			bindFilterDefs( xClass );
+		}
+	}
+
+	/**
+	 * It will add in the list all the super classes of a given class to the collection
+	 * if it's not annotated with @Indexed and it's not already in the collection.
+	 *
+	 * @param superClasses the list containing the super classes
+	 * @param mappedXClass the current class
+	 */
+	private void saveSuperClasses(Set<XClass> superClasses, XClass mappedXClass) {
+		if ( mappedXClass.getSuperclass() != null ) {
+			saveSuperClasses( superClasses, mappedXClass.getSuperclass() );
+		}
+		if ( !mappedXClass.isAnnotationPresent( Indexed.class ) && !OBJECT_CLASS_NAME.equals( mappedXClass.getName() ) ) {
```

Comment:
> Initially, I didn't want to change the existing code and I kept the patch separated. The classes annotated with @Indexed are processed later so I decided to ignore them here. I will change this so that filters are processed in one part of the code only.

Sounds good. One wonders otherwise what's going on. At least it should be commented.



>> That said, there is a ReflectionHelper.createXClassHierarchy( parseContext.getCurrentClass() ); which gives you the hierarchy as a list (bottom up). I think you could use that.

> ok

Cool

> I think I prefer the way it's done now, because the logic is going to be in one place. 

Really the metadata provider is where it should all happen and it contains the bulk of annotation processing

> but then I would have to check duplicates in SearchIntegratorBuilder anyway.

true, because the `SearchIntegrator` has the global view of all filters and analyzers. I see no contradiction here

> This way instead global elements are parsed in one place while type elements in another. I don't see it as waste unless there is reason to believe that this is going to become a performance bottleneck.

I agree, it is not a confirmed bottleneck, however, we used to be concerned about minimizing the amount of reflection calls. At least back in the days (right @emmanuelbernard !?). This will all change a bit with the introduction of Jandex, but that is still somewhere down the road.

Looking further into the code, I actually noticed that `@AnalyzerDef(s)` is also already handled in `AnnotationMetaDataProvider`. It uses `org.hibernate.search.engine.impl.ConfigContext` to keep track of the discovered analyzers definitions. The context also takes care of the de-duplication. From a configuration point of view analyzer and filter definitions are the same thing. IMO the approach should be aligned. If `ConfigContext` is used, there is even no need to do any de-duplication in `SearchIntegratorBuilder`.